app/models/user.rb

```diff
@@ -25,80 +29,165 @@ def to_s
   def set_groups_to_impersonate(grps)
-    @groups_to_impersonate = grps
+    if grps.blank? # blank is falsy values like: nil, '', []
+      # remove impersonation (see #groups below)
+      @role_cache = {}
+      @groups_to_impersonate = nil
+    elsif grps.is_a? Array
+      @role_cache = {}
+      @groups_to_impersonate = grps
+    elsif grps.is_a? String
+      @role_cache = {}
+      @groups_to_impersonate = [grps]
+    else
+      raise ArgumentError.new("Cannot accept #{grps.class} argument")
+    end
   end
 
   def groups
-    return @groups_to_impersonate if @groups_to_impersonate
-    perm_keys = ["sunetid:#{login}"]
-    return perm_keys unless webauth && webauth.privgroup.present?
-    perm_keys + webauth.privgroup.split(/\|/).collect { |g| "workgroup:#{g}" }
+    return @groups_to_impersonate unless @groups_to_impersonate.blank?
+    @groups ||= begin
+      perm_keys = ["sunetid:#{login}"]
+      return perm_keys unless webauth && webauth.privgroup.present?
+      perm_keys + webauth.privgroup.split(/\|/).map {|g| "workgroup:#{g}"}
+    end
   end
 
   # @return [Boolean] is the user a repository wide administrator
   def is_admin
-    !(groups & ADMIN_GROUPS).empty?
-  end
-
-  # @return [Boolean] is the user a repository wide viewer
-  def is_viewer
-    !(groups & VIEWER_GROUPS).empty?
+    !(groups & ADMIN_GROUPS).blank?
   end
 
   # @return [Boolean] is the user a repo wide manager
   def is_manager
-    !(groups & MANAGER_GROUPS).empty?
+    !(groups & MANAGER_GROUPS).blank?
+  end
+
+  # @return [Boolean] is the user a repository wide viewer
+  def is_viewer
+    !(groups & VIEWER_GROUPS).blank?
   end
 
   # The convention is for boolean is_XYZ methods to be interrogative (end in "?")
   alias_method :is_admin?,   :is_admin
-  alias_method :is_viewer?,  :is_viewer
   alias_method :is_manager?, :is_manager
+  alias_method :is_viewer?,  :is_viewer
 
   def login
     webauth ? webauth.login : sunetid
   end
 
-  ##
+  # Validate access permissions using Argo user roles. If that fails, check the
+  # APO role permissions.  If the object is an APO, first check permissions on
+  # the governing APO of this APO object. As a last resort, check the object
+  # roles when this object is an APO.
+  # @param dor_object [String|Dor::Base] Accepts a DRUID string or
+  #                                      any subclass of Dor::Base
+  # @return [Boolean]
+  def can_manage_object?(dor_object)
+    dor_object = parse_object(dor_object)
+    pid = dor_object.pid
+    apo_pid = dor_object.admin_policy_object.pid
+    manager_access = is_admin || is_manager
+    manager_access ||= dor_object.can_manage_item?(roles(apo_pid))
+    manager_access ||= dor_object.can_manage_content?(roles(apo_pid))
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_item?(roles(pid))
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_content?(roles(pid))
```

Comment:
The data in my `coverage/index.html` file indicates these lines are tested already.  What makes you think they are not tested or how can they be tested better?  I seem to recall the specs did hit these lines of code because there was a typo in one line (it was calling `can_manage_content` instead of `can_manage_content?` on the last line).